### DISCUSSION TITLE
Support an open source implementation of MongoDB (like FerretDB)

### DISCUSSION THREAD

**Discussion by raphj** [heyform/heyform#83] [Ideas]
Hi!

HeyForm currently depends on MongoDB 4.4, which is source-available but not open source.

I'm considering self-hosting HeyForm but non-free dependencies would be a no go. This is a suggestion to start supporting [FerretDB](https://github.com/FerretDB/FerretDB/) or another open source Mongo implementation when convenient enough.
 
FerretDB is not feature complete, and more specifically for HeyForm, its graphql endpoint fails with this error when editing the Thank you slide:

```
{"errors":[{"code":"INTERNAL_SERVER_ERROR","message":"invalid value: { \"description\": [ [ \"u\", [ [ \"b\", [ \"Thanks for\" ] ], \" completing this for\", [ \"b\", [ \"m. Now create your own form.\" ] ] ] ] ] } (nested arrays are not supported)","status":500,"stacktrace":["Error: invalid value: { \"description\": [ [ \"u\", [ [ \"b\", [ \"Thanks for\" ] ], \" completing this for\", [ \"b\", [ \"m. Now create your own form.\" ] ] ] ] ] } (nested arrays are not supported)",...,"data":null}
```

This error comes from FerretDB's `internal/types/document_validation.go` source file. Their ticket for this is https://github.com/FerretDB/FerretDB/issues/3189.

Implementing this feature in FerretDB would not be enough though, because if not specifically supported, HeyForm could break on FerretDB as soon as a new feature of MongoDB not implemented by FerretDB is used.

So, support for FerretDB or another open source implementation would be awesome.

(of course I know this is not an easy ask. Also, thanks for open sourcing HeyForm! I discovered you through [your post on HN](https://news.ycombinator.com/item?id=39895960).)

**Comment by iMuFeng** [comment]
We will test to it to check if it can be integrated in the next version.

**Reply by raphj** [reply]
Neat!

A few days ago, FerretDB [anounced support for nested arrays](https://github.com/FerretDB/FerretDB/issues/3189#issuecomment-2146580136) in the coming release.

**Reply by AlekSi** [reply]
It is now supported by FerretDB v2: https://github.com/FerretDB/FerretDB/releases/tag/v2.0.0-rc.1

**Reply by AlekSi** [reply]
@iMuFeng, hey, did you try it?

**Comment by gwozdzm** [comment]
A guide on how to use HeyForm with FerretDB is now available:
https://blog.ferretdb.io/streamlining-form-management-heyform-ferretdb/
